File: 1282-113-1408-string-matching-in-an-array/1282-113-1408-string-matching-in-an-array.js
class TrieNode {
    constructor() {
        this.children = {};
        this.freq = 0;
    }
}

var stringMatching = function(words) {
    const trie = new TrieNode();

    const insert = (word) => {
        let node = trie;
        for (const ch of word) {
            if (!node.children[ch]) {
                node.children[ch] = new TrieNode();
            }
            node = node.children[ch];
            node.freq++;
        }
    };

    const get = (word) => {
        let node = trie;
        for (const ch of word) {
            node = node.children[ch];
        }
        return node.freq > 1;
    };

    for (const word of words) {
        for (let j = 0; j < word.length; j++) {
            insert(word.substring(j));
        }
    }

    const result = [];
    for (const word of words) {
        if (get(word)) {
            result.push(word);
        }
    }

    return result;
};

File: 1282-113-1408-string-matching-in-an-array/1282-113-1408-string-matching-in-an-array.java
import java.util.*;

class TrieNode {
    TrieNode[] children = new TrieNode[26];
    int freq = 0;
}

class Solution {
    private TrieNode trie;

    public Solution() {
        this.trie = new TrieNode();
    }

    private void insert(String word) {
        TrieNode node = trie;
        for (char ch : word.toCharArray()) {
            int index = ch - 'a';
            if (node.children[index] == null) {
                node.children[index] = new TrieNode();
            }
            node = node.children[index];
            node.freq++;
        }
    }

    private boolean get(String word) {
        TrieNode node = trie;
        for (char ch : word.toCharArray()) {
            node = node.children[ch - 'a'];
        }
        return node.freq > 1;
    }

    public List<String> stringMatching(String[] words) {
        for (String word : words) {
            for (int j = 0; j < word.length(); j++) {
                insert(word.substring(j));
            }
        }

        List<String> result = new ArrayList<>();
        for (String word : words) {
            if (get(word)) {
                result.add(word);
            }
        }

        return result;
    }
}
